### ISSUE TITLE
version with neon acceleration is slower than normal ones

### ISSUE THREAD

**Issue by waterball** [FFTW/fftw3#129]
I compile fftw for android. it turns out the one with neon acceleration is somehow slower than the one without neon.

compile commands as follows:
NDK_DIR="/home/meishe01/cx/kit/android-ndk-r12b"
INSTALL_DIR="`pwd`/build-android/fftw3"

export PATH="$NDK_DIR/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/bin/:$PATH"
export SYS_ROOT="$NDK_DIR/platforms/android-16/arch-arm/"
export CC="arm-linux-androideabi-gcc --sysroot=$SYS_ROOT -march=armv7-a -mfloat-abi=softfp"
export LD="arm-linux-androideabi-ld"
export AR="arm-linux-androideabi-ar"
export RANLIB="arm-linux-androideabi-ranlib"
export STRIP="arm-linux-androideabi-strip"
export CFLAGS="-mfpu=neon -mfloat-abi=softfp"

mkdir -p $INSTALL_DIR
./configure --with-slow-timer --host=arm-linux-gnueabi --prefix=$INSTALL_DIR  LIBS="-lc -lgcc" --enable-neon --enable-float
#./configure --with-slow-timer --host=arm-linux-gnueabi --prefix=$INSTALL_DIR  LIBS="-lc -lgcc" --enable-float
make -j4
make install

I build two versions, one with neon and the other without neon, the only difference is the configure command. 

I tried both version on the same phone, meizu mx4 pro and PLK-AL10, and counted the time spent only on fftwf_execute operations(R2C and C2R). 

Any suggestions?

**Comment by waterball** [comment]
Code snippets  
 ```
 enum FFTWPlanType
    {
        planR2C = 0,
        planC2R
    };

    struct FFTWSize2D
    {
        int n0;
        int n1;

        bool operator < (const FFTWSize2D& other) const
        {
            return (n0 * n1 < other.n0 * other.n1);
        }
    };

    class FFTWPlanWrapper
    {
    public:
        FFTWPlanWrapper();
        FFTWPlanWrapper(int n0, int n1, FFTWPlanType plan_type);
        ~FFTWPlanWrapper();
        void Execute(void *in, void *out);
    private:
        fftwf_plan m_plan;
        int m_n0;
        int m_n1;
        int m_n1_padded;
        void *m_in;
        void *m_out;
        FFTWPlanType m_plan_type;
        nv::Mutex m_plan_mutex;
    };

    class FFTWPlanPool
    {
    public:
        FFTWPlanPool() {}
        ~FFTWPlanPool();

        FFTWPlanWrapper *GetPlan(int n0, int n1, FFTWPlanType plan_type);

    private:
        map<FFTWSize2D, FFTWPlanWrapper*> m_fftwf_r2c_plans;
        map<FFTWSize2D, FFTWPlanWrapper*> m_fftwf_c2r_plans;
    };
```

```
FFTWPlanWrapper::FFTWPlanWrapper(int n0, int n1, FFTWPlanType plan_type)
    : m_n0(n0)
    , m_n1(n1)
    , m_in(0)
    , m_out(0)
    , m_plan_type(plan_type)
{
    m_n1_padded = (m_n1 / 2 + 1) * 2;
    m_in = fftwf_malloc(sizeof(float) * m_n0 * m_n1_padded);
    m_out = fftwf_malloc(sizeof(float) * m_n0 * m_n1_padded);

    if (m_plan_type == planR2C)
    {
        m_plan_mutex.lock();
        m_plan = fftwf_plan_dft_r2c_2d(m_n0, m_n1, (float*)m_in, (fftwf_complex*)m_out, FFTW_ESTIMATE);
        m_plan_mutex.unlock();
    }
    else if (m_plan_type == planC2R)
    {
        m_plan_mutex.lock();
        m_plan = fftwf_plan_dft_c2r_2d(m_n0, m_n1, (fftwf_complex*)m_in, (float*)m_out, FFTW_ESTIMATE);
        m_plan_mutex.unlock();
    }
    else
        fprintf(stderr, "fftw plan type not supported!\n");

}
```

```
void FFTWPlanWrapper::Execute(void *in, void *out)
{
    double begin = nv::getTickCount();
    memcpy(m_in, in, m_n0 * m_n1_padded * sizeof(float));
    fftwf_execute(m_plan);
    memcpy(out, m_out, m_n0 * m_n1_padded * sizeof(float));
    double end = nv::getTickCount();
    double elapse = (end - begin) / nv::getTickFrequency() *1000.0;
    NvDebug() << "FFTW: " << elapse << " ms";
}
```


```
FFTWPlanWrapper::~FFTWPlanWrapper()
{
    fftwf_destroy_plan(m_plan);
    fftwf_free(m_in);
    m_in = 0;
    fftwf_free(m_out);
    m_out = 0;
}
```

```
FFTWPlanPool::~FFTWPlanPool()
{
    for (map<FFTWSize2D, FFTWPlanWrapper*>::iterator i = m_fftwf_r2c_plans.begin(); i != m_fftwf_r2c_plans.end(); ++i)
        delete i->second;
    for (map<FFTWSize2D, FFTWPlanWrapper*>::iterator i = m_fftwf_c2r_plans.begin(); i != m_fftwf_c2r_plans.end(); ++i)
        delete i->second;
    fftwf_cleanup();
}
```

```
FFTWPlanWrapper *FFTWPlanPool::GetPlan(int n0, int n1, FFTWPlanType plan_type)
{
    FFTWSize2D size = {n0, n1};

    FFTWPlanWrapper *plan;
    if (plan_type == planR2C)
    {
        if (m_fftwf_r2c_plans.find(size) == m_fftwf_r2c_plans.end())
        {
            plan = new FFTWPlanWrapper(n0, n1, plan_type);
            m_fftwf_r2c_plans[size] = plan;
            return plan;
        }
        else
        {
            return m_fftwf_r2c_plans.find(size)->second;
        }
    }
    else if (plan_type == planC2R)
    {
        if (m_fftwf_c2r_plans.find(size) == m_fftwf_c2r_plans.end())
        {
            plan = new FFTWPlanWrapper(n0, n1, plan_type);
            m_fftwf_c2r_plans[size] = plan;
            return plan;
        }
        else
        {
            return m_fftwf_c2r_plans.find(size)->second;
        }
    }
    else
    {
        fprintf(stderr, "fftw plan type not supported!\n");
        return nullptr;
    }
}
```

**Comment by waterball** [comment]
Forget the code above. I've written a speed testing sample in Android Studio for R2C operations. At the same time, I've debuged into the simd_neon.h files so that I'm pretty sure that fftw uses neon to optimize fft. Sample code is [here,](https://github.com/waterball/FFTWTest/tree/master) Code snippets as:
```
    const int size = 32;
    float *in = (float *)fftwf_malloc(size * size * sizeof(float));
    fftwf_complex *out = (fftwf_complex *)fftwf_malloc(size * (size+2) * sizeof(float));
    fftwf_plan p = fftwf_plan_dft_r2c_2d(size, size, in, out, FFTW_ESTIMATE);

    timeval begin, end;
    double elapse;
    gettimeofday(&begin, 0);
    for (int i = 0; i < 1000; ++i)
        fftwf_execute(p);
    gettimeofday(&end, 0);
    elapse = 1000.0 * (end.tv_sec - begin.tv_sec) + (end.tv_usec - begin.tv_usec) / 1000.0;
    elapse = elapse / 1000.0;

    char elapse_s[100];
    sprintf(elapse_s, "Elapse: %f ms\n", elapse);
    fftwf_destroy_plan(p);
    fftwf_free(in);
    fftwf_free(out); 
```
I tested R2C operation on Meizu mx4, and came to a wierd result as follows:
 

size | 32 | 64 | 80 | 128
-- | -- | -- | -- | --
with neon | 0.053 ms | 0.38 ms | 1.19 ms | 2.8 ms
without neon | 0.041 ms | 0.46 ms | 0.62 ms | 2.66 ms

We see that in most cases fftw without neon is faster than the other.
If the way I use FFTW is wrong, correct me. Thanks!!!

**Comment by NPellet** [comment]
Actually I'm a bit confused about the -mfpu option.
In configure.ac, the following appears

```
        case "${host_cpu}" in
            aarch64)
                ;;
            *)
                if test "$have_neon" = "yes" -a "x$NEON_CFLAGS" = x; then
                    AX_CHECK_COMPILER_FLAGS(-mfpu=neon, [NEON_CFLAGS="-mfpu=neon"],
                [AC_MSG_ERROR([Need a version of gcc with -mfpu=neon])])
                fi
                ;;
        esac
```

But the aarch64 reference states that the -mfpu flag is ignored ("-mfpu=list is rejected when targeting AArch64.", see https://developer.arm.com/documentation/100067/0608/armclang-Command-line-Options/-mfpu?lang=en)

The documention goes on to state that the -mcpu option is the relevant one for aarch64.

**Comment by Cryolitia** [comment]
We tested and compared the benchmark of fftw 3.3.10 on NixOS as a part of daily maintaining, here's the result: https://github.com/NixOS/nixpkgs/pull/484649

TLDR: on most arm64 devices we have, build fftw with arm neon support resulting in amount 10% faster
